File: asm-lab/src/smoke.c
/* asm-lab smoke test — validates the toolchain + jsnes memory plumbing.
 *
 * Result buffer is internal RAM at $0300 (the harness reads it via
 * nes.cpu.mem[0x0300+]). We prove the $6000 WRAM round-trip by writing a
 * marker to a volatile WRAM global and reading it back into the buffer:
 *   $0300 = 0x43, $0301 = 0xA5   (direct internal-RAM writes)
 *   $0302 = g_wram               (== 0x42 iff jsnes emulates $6000 WRAM)
 * volatile forces -Os to emit the real store + load instead of folding it.
 */
volatile unsigned char g_wram;   /* -> BSS ($6000 WRAM) */

void main(void) {
    volatile unsigned char *buf = (unsigned char *)0x0300;
    g_wram = 0x42;
    buf[0] = 0x43;
    buf[1] = 0xA5;
    buf[2] = g_wram;   /* real load from $6000 */
    for (;;) { }
}

File: asm-lab/harness/smoke.mjs
// Smoke test: boot build/smoke.nes, run a few frames, verify the internal-RAM
// buffer and the $6000 WRAM round-trip.
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { boot, makeReporter } from './nes.mjs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const r = makeReporter('asm-lab smoke');
const h = boot(path.join(__dirname, '..', 'build', 'smoke.nes'));
h.frames(4);

r.eq('$0300 (direct write)', h.rd(0x0300), 0x43);
r.eq('$0301 (direct write)', h.rd(0x0301), 0xA5);
r.eq('$0302 (read back from $6000 WRAM global)', h.rd(0x0302), 0x42);
r.done('asm-lab smoke: toolchain + jsnes RAM/WRAM plumbing OK.');
